Content.Server/Nutrition/EntitySystems/VapeSystem.cs

```diff
@@ -0,0 +1,122 @@
+
+using Content.Server.Nutrition.Components;
+using Content.Shared.Chemistry.Components;
+using Content.Server.Body.Components;
+using Content.Shared.Interaction;
+using Content.Server.DoAfter;
+using System.Threading;
+using Content.Server.Body.Systems;
+using Content.Server.Chemistry.EntitySystems;
+using Content.Server.Explosion.EntitySystems;
+
+namespace Content.Server.Nutrition.EntitySystems
+{
+    public sealed class VapeSystem : EntitySystem
+    {
+        [Dependency] private readonly DoAfterSystem _doAfterSystem = default!;
+        [Dependency] private readonly SolutionContainerSystem _solutionContainerSystem = default!;
+        [Dependency] private readonly BloodstreamSystem _bloodstreamSystem = default!;
+        [Dependency] private readonly FoodSystem _foodSystem = default!;
+        [Dependency] private readonly ExplosionSystem _explosionSystem = default!;
+
+        public override void Initialize()
+        {
+            base.Initialize();
+            SubscribeLocalEvent<VapeComponent, AfterInteractEvent>(OnInteraction);
+
+            SubscribeLocalEvent<VapingEvent>(InjectAndSmokeReagents);
+        }
+
+        private void OnInteraction(EntityUid uid, VapeComponent comp, AfterInteractEvent args) 
+        {
+            _solutionContainerSystem.TryGetRefillableSolution(uid, out var solution);
+
+            var delay = comp.Delay;
+
+            if(!TryComp<BloodstreamComponent>(args.Target, out var bloodstream)
+            || solution == null
+            || !args.CanReach
+            || _foodSystem.IsMouthBlocked(args.Target.Value, args.User))
+                return;
+
+            if(args.Target == args.User)
+                delay = comp.UserDelay;
+            
+            var entMan = IoCManager.Resolve<IEntityManager>();
+
+            //Always read the description.
+            if(solution.ContainsReagent("Water") || comp.ExplodeOnUse)
+            {
+                _explosionSystem.QueueExplosion(uid, "Default", comp.ExplosionIntensity, 0.5f, 3, canCreateVacuum: false);
+                entMan.DeleteEntity(uid);
+            }
+
+            comp.CancelToken = new CancellationTokenSource();
+            _doAfterSystem.DoAfter(new DoAfterEventArgs(args.User, delay, comp.CancelToken.Token, args.Target)
+            {
+                BreakOnTargetMove = true,
+                BreakOnUserMove = false,
+                BreakOnDamage = true,
+                BreakOnStun = true,
+                BroadcastFinishedEvent = new VapingEvent(args.User, args.Target.Value, uid, comp, solution, bloodstream)
```

Comment:
just used a `UsedFinishedEvent` so that you can just do a normal event subscription in `Initialize()`.